beacon_node/beacon_chain/src/fetch_blobs.rs

```diff
@@ -125,136 +164,160 @@ pub async fn fetch_and_process_engine_blobs<T: BeaconChainTypes>(
+            block.slot(),
+            block_root,
+            EngineGetBlobsOutput::Blobs(fixed_blob_sidecar_list.clone()),
+        )
+        .await
+        .map_err(FetchEngineBlobError::BlobProcessingError)?;
 
-        if chain
-            .canonical_head
-            .fork_choice_read_lock()
-            .contains_block(&block_root)
-        {
-            // Avoid computing columns if block has already been imported.
-            debug!(
-                info = "block has already been imported",
-                "Ignoring EL blobs response"
-            );
-            return Ok(None);
-        }
+    Ok(Some(availability_processing_status))
+}
 
-        let data_columns_receiver = spawn_compute_and_publish_data_columns_task(
-            &chain,
-            block.clone(),
-            fixed_blob_sidecar_list.clone(),
-            publish_fn,
-        );
+async fn fetch_and_process_blobs_v2<T: BeaconChainTypes>(
+    chain: Arc<BeaconChain<T>>,
+    block_root: Hash256,
+    block: Arc<SignedBeaconBlock<T::EthSpec>>,
+    versioned_hashes: Vec<VersionedHash>,
+    custody_columns_indices: HashSet<ColumnIndex>,
+    publish_fn: impl Fn(BlobsOrDataColumns<T>) + Send + 'static,
+) -> Result<Option<AvailabilityProcessingStatus>, FetchEngineBlobError> {
+    let num_expected_blobs = versioned_hashes.len();
+    let execution_layer = chain
+        .execution_layer
+        .as_ref()
+        .ok_or(FetchEngineBlobError::ExecutionLayerMissing)?;
+
+    metrics::observe(&metrics::BLOBS_FROM_EL_EXPECTED, num_expected_blobs as f64);
+    debug!(num_expected_blobs, "Fetching blobs from the EL");
+    let response = execution_layer
+        .get_blobs_v2(versioned_hashes)
+        .await
+        .inspect_err(|_| {
+            inc_counter(&metrics::BLOBS_FROM_EL_ERROR_TOTAL);
+        })
+        .map_err(FetchEngineBlobError::RequestFailed)?;
+
+    let (blobs, proofs): (Vec<_>, Vec<_>) = response
+        .into_iter()
+        .filter_map(|blob_and_proof_opt| {
+            blob_and_proof_opt.map(|blob_and_proof| {
+                let BlobAndProofV2 { blob, proofs } = blob_and_proof;
+                (blob, proofs)
+            })
+        })
+        .unzip();
 
-        Some(data_columns_receiver)
+    let num_fetched_blobs = blobs.len();
+    metrics::observe(&metrics::BLOBS_FROM_EL_RECEIVED, num_fetched_blobs as f64);
+
+    if num_fetched_blobs == 0 {
+        debug!(num_expected_blobs, "No blobs fetched from the EL");
+        inc_counter(&metrics::BLOBS_FROM_EL_MISS_TOTAL);
+        return Ok(None);
     } else {
-        if !blobs_to_import_and_publish.is_empty() {
-            publish_fn(BlobsOrDataColumns::Blobs(blobs_to_import_and_publish));
-        }
+        inc_counter(&metrics::BLOBS_FROM_EL_HIT_TOTAL);
+    }
 
-        None
-    };
+    // Partial blobs response isn't useful for PeerDAS, so we don't bother building and publishing data columns.
+    if num_fetched_blobs != num_expected_blobs {
```

Comment:
Above you check `num_fetched_blobs == 0`. Can `num_expected_blobs` be 0? Else you can remove the above condition. Otherwise add here the `inc_counter(&metrics::BLOBS_FROM_EL_MISS_TOTAL);` as this is also a miss.

Reply:
if `num_expected_blobs` is 0, it's returned early here
https://github.com/sigp/lighthouse/blob/73fe3175bd30c9ead1f66593435be56e76de5c9e/beacon_node/beacon_chain/src/fetch_blobs.rs#L78-L79

`BLOBS_FROM_EL_MISS_TOTAL` was originally used for tracking empty responses, but we can repurpose that. I generally avoid making these unrelated changes to keep the PR concise and easier to review, but happy to do it since you're reviewing :P
https://github.com/sigp/lighthouse/blob/70850fe58d5676d9072d418f04f4db7257c1513b/beacon_node/beacon_chain/src/metrics.rs#L1663

I think this file can use some refactoring, but I have also avoided doing this for the interest of keeping this moving quickly.